### PR TITLE
Improve test coverage to 80%

### PR DESCRIPTION
**Pull request by hamishmorgan**
Test coverage was poor. This PR adds a bunch of unit tests to improve things somewhat. I also deleted any unused code I came across.

Based on https://github.com/BrandwatchLtd/robots/pull/19

### CODE REVIEW

File: core/src/test/java/com/brandwatch/robots/RobotsConfigTest.java [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void givenNegativeValue_whenSetCacheExpiresHouse_thenThrowsIAE() {
+        final long invalidCacheExpiresHours = -1;
+        config.setCacheExpiresHours(invalidCacheExpiresHours);
+    }
+
+    @Test
+    public void givenZeroValue_whenSetCacheExpiresHouse_thenValueIsStored() {
```
Comment (jstanier) [inline core/src/test/java/com/brandwatch/robots/RobotsConfigTest.java]: Typo: house. Same as on other tests in this class too.

**Comment by jstanier** [comment]
Heroic effort. Found a couple of typos. Nothing to add apart from that!

**Comment by hamishmorgan** [comment]
Well spotted, thanks. 

I was a little disappointed with a messily 9% improvement after all that. I'm aiming for 80%. There are a number of areas that are hard to unit test, like the generated parser code for example.

**Comment by jstanier** [comment]
Almost into the yellow zone of 80%+!

**Comment by hamishmorgan** [comment]
The core module is well over 90% already. I'm having difficulty with the JavaCC generated code, which contains a whole bunch of unused code. This leaves me with a choice of either writing tests for irrelevant implementation details, or removing the generated code from the coverage report. Neither seems like a great plan.

**Comment by hamishmorgan** [comment]
Yellow is now my favourite colour.

**Comment by spmason** [comment]
You've done bloody well to get this far! 

But (having not read much into the code on this pr) I'd ask how much it really matters? Are you any more confident in the code after all this? If all the critical paths are covered and you're pretty happy in the code being stable then I wouldn't chase coverage figures much more.

I've always gone for 60-70% before so 80 with auto-generated code seems pretty good going!

**Comment by hamishmorgan** [comment]
I've certainly been thinking about whether this is worth the effort. A ~~pessimistic~~ realistic view is that coverage puts an upper bound on correctness (80% coverage implies 20% is completely untested, not that the 80% is correct). 

Having said that, I _did_ uncover a couple of unexpected "behaviours" while writing these tests. My previous moaning in the comments, about difficulty testing generated code, was because I had overlooked testing various non-printable and non-ASCII code-points. And there in lies the point coverage; it directs you to untested areas of the code-base.

So yes, I set myself an arbitrary target, and yes, I could have got there without testing anything. You just have to decide if my love of the new shiny yellow badge is greater than my love of correct code. :smiley:

File: core/src/test/java/com/brandwatch/robots/RobotsServiceImplTest.java [hunk 2]
```diff
@@ -38,76 +38,135 @@
...
-        String crawlerAgent = "magpie";
-        URI resourceUri = URI.create("http://example.org/index.html");
-        when(utilities.getRobotsURIForResource(resourceUri)).thenThrow(new IllegalArgumentException());
-        instance.isAllowed(crawlerAgent, resourceUri);
+        when(utilities.getRobotsURIForResource(RESOURCE_URI)).thenThrow(new IllegalArgumentException());
+        instance.isAllowed(CRAWLER_AGENT, RESOURCE_URI);
+    }
+
+    @Test
+    public void whenClose_thenCloseCalledLoadere() throws IOException {
```
Comment (jstanier) [inline core/src/test/java/com/brandwatch/robots/RobotsServiceImplTest.java]: Typo

**Comment by jstanier** [comment]
Found one typo. Amazing effort! Will merge it when that's fixed.

**Comment by hamishmorgan** [comment]
Thanks @jstanier - I've fixed the typo and re-based.

**Pull request merged**